prow/cmd/peribolos/main.go

```diff
@@ -103,21 +128,132 @@ func main() {
+	for _, r := range opt.requiredAdmins.Strings() {
+		if !wantAdmins.Has(r) {
+			missing = append(missing, r)
+		}
+	}
+	if len(missing) > 0 {
+		return fmt.Errorf("%s must specify %v as admins, missing %v", orgName, opt.requiredAdmins, missing)
+	}
+	if opt.requireSelf {
+		if me, err := client.BotName(); err != nil {
+			return fmt.Errorf("cannot determine user making requests for %s: %v", opt.token, err)
+		} else if !wantAdmins.Has(me) {
+			return fmt.Errorf("authenticated user %s is not an admin of %s", me, orgName)
+		}
+	}
+
+	// Get current state
+	haveAdmins := sets.String{}
+	haveMembers := sets.String{}
+	if ms, err := client.ListOrgMembers(orgName, "admin"); err != nil {
+		return fmt.Errorf("failed to list %s admins: %v", orgName, err)
+	} else {
+		for _, m := range ms {
+			haveAdmins.Insert(m.Login)
+		}
+	}
+
+	if ms, err := client.ListOrgMembers(orgName, "member"); err != nil {
+		return fmt.Errorf("failed to list %s members: %v", orgName, err)
+	} else {
+		for _, m := range ms {
+			haveMembers.Insert(m.Login)
+		}
+	}
+
+	// Ensure desired state is sane
+	both := wantAdmins.Intersection(wantMembers)
+	if n := len(both); n > 0 {
+		return fmt.Errorf("%d users are both a member and admin: %v", n, both)
+	}
+
+	// Figure out who to remove
+	have := haveMembers.Union(haveAdmins)
+	want := wantMembers.Union(wantAdmins)
+	remove := have.Difference(want)
+
+	// Figure out who to invite and/or reconfigure
+	makeMember := wantMembers.Difference(haveMembers)
+	makeAdmin := wantAdmins.Difference(haveAdmins)
+
+	// Sanity check changes
+	if d := float64(len(remove)) / float64(len(have)); d > opt.maximumDelta {
+		return fmt.Errorf("cannot delete %d memberships or %.3f of %s (exceeds limit of %.3f)", len(remove), d, orgName, opt.maximumDelta)
+	}
+
+	var errs []error
+
+	// March towards desired state
+	for u := range remove {
+		if err := client.RemoveOrgMembership(orgName, u); err != nil {
+			logrus.WithError(err).Warnf("RemoveOrgMembership(%s, %s) failed", orgName, u)
+			errs = append(errs, err)
+		} else {
+			logrus.Infof("Removed %s from %s", u, orgName)
+		}
+	}
+
+	for u := range makeMember {
+		if m, err := client.UpdateOrgMembership(orgName, u, false); err != nil {
+			logrus.WithError(err).Warnf("UpdateOrgMembership(%s, %s, false) failed", orgName, u)
+			errs = append(errs, err)
+		} else if m.State == github.StatePending {
+			logrus.Infof("Invited %s to %s", u, orgName)
+		} else {
+			logrus.Infof("Demoted %s to a %s member", u, orgName)
+		}
+	}
+
+	for u := range makeAdmin {
+		if m, err := client.UpdateOrgMembership(orgName, u, true); err != nil {
+			logrus.WithError(err).Warnf("UpdateOrgMembership(%s, %s, true) failed", orgName, u)
+			errs = append(errs, err)
+		} else if m.State == github.StatePending {
+			logrus.Infof("Invited %s to %s", u, orgName)
+		} else {
+			logrus.Infof("Promoted %s to a %s admin", u, orgName)
+		}
+	}
+
+	if n := len(errs); n > 0 {
+		return fmt.Errorf("%d errors: %v", n, errs)
```

Comment:
You're going to inform the user of every error twice this way -- once through `logrus` at run-time and once down here at the end.

Reply:
This one is going to be a non-formatted glob mess of text, this is also how label sync does it and FWIW I find the dump at the end of the run to be unintelligible